### DISCUSSION TITLE
whether support debug project?

### DISCUSSION THREAD

**Discussion by changpenghui616** [memtest86plus/memtest86plus#106] [Q&A]
As README.md describe，project build on linux system，and run memtest.efi on machine
I try to understand the memtest logic ， so whether has some guide to debug code？

**Comment by changpenghui616** [comment]
i try use file_protocol in project.
in efi_setup.c can work, 
but in main.c，it cant work, and display "Unexpected interrupt on cpu", i guess pointer error generate progrem crash when use ‘locate_handle’,but dont know what happend
code：
~~~
    boot_params_t *boot_params = (boot_params_t *)boot_params_addr;
    efi_system_table_t *sys_table = NULL;
    uintptr_t sys_table_temp = (uintptr_t)boot_params->efi_info.sys_tab;
    sys_table = (efi_system_table_t*)sys_table_temp;
    if (sys_table == NULL || boot_params == NULL) {
        return false;
    }
    status = efi_call_bs(locate_handle, EFI_LOCATE_BY_PROTOCOL, &gEfiSimpleFileSystemProtocolGuid, NULL, &handles_size, handles);
    if (status == EFI_BUFFER_TOO_SMALL) {
        return false;
    }
~~~

**Comment by martinwhitaker** [comment]
`efi_setup()` calls `EFI_BOOT_SERVICES.ExitBootServices()`. No UEFI boot services are available after that. This is essential because Memtest86+ does not preserve the UEFI environment.

**Reply by changpenghui616** [reply]
i find at main.c global_init()/hwctrl_init() has used uefi runtime_services, i think if can get boot_service pointer,just like runtime_services, whether can work?
i see, EFI_BOOT_SERVICES.ExitBootServices() has desotry boot_services, so cant use this pointer

**Reply by changpenghui616** [reply]
whether can move EFI_BOOT_SERVICES.ExitBootServices() to program end? what happend if do this

**Reply by martinwhitaker** [reply]
As I said, Memtest86+ doesn't preserve the UEFI environment. It overwrites the memory used by the UEFI boot services with test patterns (because we want to test that memory). And it changes the virtual memory map. So the UEFI boot services would very soon stop working, even if you delayed calling `ExitBootServices()`.

**Reply by changpenghui616** [reply]
Thanks help.
As you say，it can't run and display black screen ，when i try to delayed calling ExitBootServices().
I wonder to know whether have  plan to create and save report or log about test result when memtest86 completely（pass/fail）?
ps：i find a interface（serial_echo_print（））in serical.c，it seem to output message to serial port，but when i try to catch serial port log，it get nothing.

**Reply by debrouxl** [reply]
Saving report to persistent storage, be it USB MSD or internal HDD / SSD, with e.g. a FAT32 partition, is a known feature request, indeed.
However, many computers on which running memtest86+ remains relevant don't support UEFI boot at all, or their users disable it even on modern computers for whichever reason. Therefore, IMO (others may beg to differ), the persistent storage report writing functionality is only good if it supports both legacy BIOS and UEFI.
If and when memtest86+ aims at writing reports to persistent storage, it may just be better to rewrite memtest86+ as a task in an existing RTOS which already has drivers for USB MSD, internal IDE + SATA + NVMe (+ SAS) controllers and devices, FAT32 and possibly other filesystems. IMO, again :)

**Reply by changpenghui616** [reply]
As you say,if rewrite memtest86 as RTMOS task, whether need a lot of code，and some memory be use?
I used memtest86（[https://www.memtest86.com/） once，it can save report to FAT32 when run completely. So i think add this feature in current memtest86+ framework is choice better，if can realize.
After all， memtest86+ one of new feature is support UEFI boot :)

**Reply by martinwhitaker** [reply]
Memtest86 was rewritten to be a UEFI application. That means that:

1. It doesn't work on machines with a legacy BIOS.
2. It can't test the memory being used by the UEFI BIOS.

On my machines, the UEFI BIOS uses between 100MB and 200MB of memory.

So I chose not to go down that path.

**Reply by x86fr** [reply]
And that was the best thing to do. Some newer, full implementation of the TianoCore EDK II now requires 500MB in memory.

We won't need support for NVMe, IDE, SATA or whatever. But at some point, being able to write on the FAT32-formatted USB Key we're using to run Memtest86+ could be interesting for many users.

**Reply by changpenghui616** [reply]
I see，when memtest86+ run，it exit the UEFI environment already（run at cache, maybe UEFI just support for boot machine and init test environment），but memtest86 run at UEFI environment all time（run at memory，so cant test all memory space），this is difference between the two framework？I dont know whether understand right.
Finally，I still want to know whether have the develop plan that save test report about memtest86+，maybe support this feature in In the future（such as RTMOS, ..）？I think this is important for users to  analyze memory problems with high-efficiency

**Reply by martinwhitaker** [reply]
Your question about the development plan for saving test reports was already answered by @debrouxl and @x86fr earlier in this thread. It is hard to implement, so won't happen soon.

**Reply by changpenghui616** [reply]
okay,thanks help
